### PR TITLE
Match Feedback Forms: Modified overall logic based on updated API spec

### PR DESCRIPTION
**Pull request by noah-solomon**
## Overview

Due to design changes, some of the fields in our Match Feedback model have changed, as well as the logic to interact with those fields. For the most up-to-date API specification, search "SP22 Pear Django API Specification" on Notion.

## Changes Made

- Modified URL structure to make it clear that `Feedback` falls within the realm of a `Match`
- Fixed bug for returned response in `CreateMatchController` when single matches were created
- Allow users to directly set their own `pending_feedback` flag if they close out of the form
- Added condensed storing of feedback options when a user does NOT meet up with their match to increase space efficiency, and added utility methods to interact with these condensed options
- Automatically set `pending_feedback` to `True` when new matches are created
- Modified overall Feedback creation process due to change in designs and in API spec

## Test Coverage

Manual postman testing

### CODE REVIEW

File: src/match/controllers/create_match_controller.py [hunk 1]
```diff
@@ -16,31 +16,39 @@ def process(self):
...
             error_msg = f"User with id {min(match_ids)} does not exist"
-            return False, error_msg if self._return_status else failure_response(
-                error_msg
+            return (
+                (False, error_msg)
+                if self._return_status
+                else failure_response(error_msg)
             )
         # Always set user_2 to the higher user id
         user_2 = User.objects.filter(id=max(match_ids))
```
Comment (alanna-zhou) [inline src/match/controllers/create_match_controller.py:34]: doesnt this... give you a list? how are you even accessing user_2 rn..

`User.objects.filter(id=max(match_ids)).exists()` is what you're looking for? it's a `bool`

File: src/match/controllers/create_match_controller.py [hunk 2]
```diff
@@ -16,31 +16,39 @@ def process(self):
         match_ids = self._data.get("ids")
         if match_ids is None:
             error_msg = "POST body is misformatted"
-            return False, error_msg if self._return_status else failure_response(
-                error_msg, status.HTTP_400_BAD_REQUEST
+            return (
+                (False, error_msg)
+                if self._return_status
+                else failure_response(error_msg, status.HTTP_400_BAD_REQUEST)
```
Comment (alanna-zhou) [inline src/match/controllers/create_match_controller.py:22]: this is a bit strange

why can't we just return directly 
`failure_response(error_msg, self._return_status or status.HTTP_400_BAD_REQUEST)`

File: src/match/controllers/create_match_controller.py [hunk 3]
```diff
@@ -51,6 +59,10 @@ def process(self):
...
         match.proposed_locations.set([])
         match.save()
-        return True, "" if self._return_status else success_response(
-            None, status.HTTP_201_CREATED
+        user_1.person.pending_feedback = True
+        user_2.person.pending_feedback = True
+        return (
+            (True, "")
+            if self._return_status
+            else success_response(None, status.HTTP_201_CREATED)
```
Comment (alanna-zhou) [inline src/match/controllers/create_match_controller.py:67]: this and all the others same comment as my first comment

File: src/survey/constants.py [hunk 4]
```diff
@@ -6,3 +6,45 @@
 STRONGLY_GOOD = 5
 
 RATINGS = [STRONGLY_BAD, SLIGHTLY_BAD, NEUTRAL, SLIGHTLY_GOOD, STRONGLY_GOOD]
+
+BAD_FIT = ("BF", "Not a good fit")
+NO_RESPONSE = ("NR", "They didn't respond")
```
Comment (alanna-zhou) [inline src/survey/constants.py:11]: let's keep grammar consistent -- `Did not respond`

File: src/survey/constants.py [hunk 5]
```diff
@@ -6,3 +6,45 @@
...
+    elif long_name == NO_RESPONSE[1]:
+        return NO_RESPONSE[0]
+    elif long_name == NOT_INTERESTED[1]:
+        return NOT_INTERESTED[0]
+    elif long_name == TOO_BUSY[1]:
+        return TOO_BUSY[0]
+    elif long_name == OTHER[1]:
+        return OTHER[0]
+    else:
+        return None
```
Comment (alanna-zhou) [inline src/survey/constants.py:50]: what is the point of converting from long to short name
why can't it just be a dictionary mapping too, then it can just return the `value` given the `key`

File: src/survey/controllers/create_survey_controller.py [hunk 6]
```diff
@@ -1,49 +1,75 @@
...
 
 # TODO(@chalo2000) Convert to celery task
 class CreateSurveyController:
-    def __init__(self, request, data):
+    def __init__(self, request, data, match_id):
         self._request = request
         self._data = data
+        self._match_id = match_id
 
     def process(self):
```
Comment (alanna-zhou) [inline src/survey/controllers/create_survey_controller.py:20]: im glad you guys are using controllers and using `process` :))) this is an instagram coding standard <3

File: src/survey/controllers/create_survey_controller.py [hunk 7]
```diff
@@ -1,49 +1,75 @@
...
-            )
+            return failure_response("match_id is invalid", status.HTTP_404_NOT_FOUND)
         completed_match = completed_match[0]
 
         # Check if the submitting user has already submitted a survey
-        submitting_user = Person.objects.filter(id=self._request.user.id)[0]
+        submitting_person = self._request.user.person
         survey = Survey.objects.filter(
-            submitting_user=submitting_user, completed_match=completed_match
+            submitting_person=submitting_person, completed_match=completed_match
```
Comment (alanna-zhou) [inline src/survey/controllers/create_survey_controller.py:72]: this looks a bit convoluted

this might be better: `Survey.objects.filter(submitting_person=self._request.user.person)`

remember that you can always do `Survey.objects.filter(submitting_user__person=self._request.user.person)` but here it's not necessary to do that bc you guys just did a weird naming thing

File: src/survey/serializers.py [hunk 8]
```diff
@@ -1,9 +1,32 @@
...
 
 class SurveySerializer(serializers.ModelSerializer):
+
+    did_not_meet_reasons = SerializerMethodField("get_did_not_meet_reasons")
+
+    def get_did_not_meet_reasons(self, survey):
+        if survey.did_not_meet_reasons is None:
+            return []
+        return map(
+            lambda x: constants.short_to_long(x),
```
Comment (alanna-zhou) [inline src/survey/serializers.py:17]: if you do a dictionary like i mentioned, you can just return `dict.keys()` or `dict.values()` or something easier like that instead of using convoluted lambda funcs

File: src/survey/views.py [hunk 9]
```diff
@@ -32,11 +33,31 @@ def post(self, request):
...
+
+
 class SurveyView(generics.GenericAPIView):
     serializer_class = SurveySerializer
     permission_classes = api_settings.CONSUMER_PERMISSIONS
 
-    def get(self, request, id):
+    def get(self, request, id, match_id):
         """Get survey by id."""
         survey = Survey.objects.filter(id=id)
```
Comment (alanna-zhou) [inline src/survey/views.py:62]: this returns a list...

Comment (alanna-zhou) [reply src/survey/views.py:62]: i am very confused by the whole checking system here

if you want to returns omething by id, the best solution is just to give

```
if Survey.objects.filter(id=id).exists():
  return success_response ...(serializer... (Survey.objects.get(id=id))...
return failure_response ...
```

**Review by alanna-zhou — CHANGES_REQUESTED** [review]
nice job!! this is a pretty massive pr, lots of good changes!

**Comment by alanna-zhou** [comment]
always tag me in PRs from now on!

**Review by jzs27 — APPROVED** [review]

File: src/survey/constants.py (hunk 5 above)
Comment (chalo2000) [reply src/survey/constants.py:50]: These functions could've been simplified with a loop checking DID_NOT_MEET rather than the tuple values. I agree with @alanna-zhou  about the dictionary, just make one and make the other by reversing the keys and values and choose the appropriate dictionary when converting.

File: src/survey/controllers/create_survey_controller.py [hunk 10]
```diff
@@ -1,49 +1,75 @@
...
-            return failure_response(
-                "POST body is misformatted", status.HTTP_400_BAD_REQUEST
-            )
+        if did_meet is None:
+            return failure_response("did_meet", status.HTTP_400_BAD_REQUEST)
 
-        # Get optional fields
+        # Get optional fields based on did_meet
+        did_meet_reason = self._data.get("did_meet_reason")
+        did_not_meet_reasons = self._data.get("did_not_meet_reasons")
```
Comment (chalo2000) [inline src/survey/controllers/create_survey_controller.py:28]: You can get `did_not_meet_reasons` and `rating` inside the if statements since they're only used in specific conditions.

Comment (chalo2000) [reply src/survey/controllers/create_survey_controller.py:28]: nevermind I see them being used at the bottom of this

File: src/survey/views.py [hunk 11]
```diff
@@ -12,8 +12,9 @@
 from .serializers import SurveySerializer
 
 
-class SurveysView(generics.GenericAPIView):
+class AllSurveysView(generics.GenericAPIView):
     serializer_class = SurveySerializer
+    # TODO @njs99: figure out the best way to limit this to superusers
```
Comment (chalo2000) [inline src/survey/views.py:17]: wouldn't all you have to do is set the permission classes to `api_settings.ADMIN_PERMISSIONS` ?

File: src/survey/views.py (hunk 9 above)
Comment (chalo2000) [reply src/survey/views.py:62]: This way is actually pretty clean and it's clear what is being done, I think we can do a refactor in the future to do it like this - you can ignore my earlier comment where you should check the length to make it more explicit and adopt this style.

File: src/match/controllers/create_match_controller.py (hunk 1 above)
Comment (chalo2000) [reply src/match/controllers/create_match_controller.py:34]: Actually scroll down to my later comment, @alanna-zhou suggested a cleaner, clearer way of doing this.

File: src/survey/views.py [hunk 12]
```diff
@@ -45,15 +66,15 @@ def get(self, request, id):
             self.serializer_class(survey[0]).data, status.HTTP_200_OK
         )
 
-    def post(self, request, id):
```
Comment (chalo2000) [inline src/survey/views.py:48]: Are we going to make use of these `match_id` parameters for `post` and `delete` in this PR? If not, make this change in the PR where you'll actually make use of them.

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Let's get some changes in before I approve this.

File: src/survey/serializers.py (hunk 8 above)
Comment (noah-solomon) [reply src/survey/serializers.py:17]: when @alanna-zhou  doesn't like my lambda functions 😢

Comment (noah-solomon) [reply src/survey/serializers.py:17]: yes good call

File: src/survey/views.py (hunk 11 above)
Comment (noah-solomon) [reply src/survey/views.py:17]: yea u right lol, i thought i had ran into issues with that before but I just double-checked and its working on `/api/matches/` for something similar

File: src/survey/views.py (hunk 12 above)
Comment (noah-solomon) [reply src/survey/views.py:48]: `match_id` is irrelevant but it's in the URL structure to maintain consistency with the other endpoints (i.e. creating a survey, for which you do need `match_id`). Because it's in the URL structure, it has to be in the method header because otherwise Django will try to pass it in and Python will get mad
tl;dr: i removed the parameters i didn't need and it raised an error

File: src/match/controllers/create_match_controller.py (hunk 2 above)
Comment (noah-solomon) [reply src/match/controllers/create_match_controller.py:22]: yeah this part was created by @chalo2000 to work with the matching algorithm - I had to add parenthesis because it didn't work how he intended before. there's probably cleaner ways to do it but then we would have to change the matcher submodule too so I'm gonna leave it for now

File: src/survey/constants.py (hunk 5 above)
Comment (noah-solomon) [reply src/survey/constants.py:50]: good call

File: src/survey/serializers.py (hunk 8 above)
Comment (noah-solomon) [reply src/survey/serializers.py:17]: uhh wait I still need to use a lambda bc i'm converting from keys to values (or the other direction), let me know if you had another solution in mind

File: src/survey/views.py (hunk 9 above)
Comment (noah-solomon) [reply src/survey/views.py:62]: the code I have here does the same thing but yeah I can change for readability purposes

**Review by alanna-zhou — APPROVED** [review]
u da GOAT 😫😫😫
